### PR TITLE
Fix scrollbar auto-show for cursors out of sight

### PR DESCRIPTION
**Pull request by alygin**
This PR fixes scrollbar auto-show feature when there're not visible cursors. The intial behavior was broken in #11147.

The problem is that `selections` only contains visible selections, so the `if` with `non_visible_cursors |= true` is only visited in rare edge cases when we have at least part of the selection still visible on the screen. But when we scroll far enough from the cursor, `non_visible_cursors` keeps its default `false` value, which is incorrect.

Release Notes:

- N/A

/cc @mikayla-maki

### CODE REVIEW

**Review by SomeoneToIgnore — CHANGES_REQUESTED** [review]
There's something fishy with the way this feature works: the original revert was due to the fact that the entire scroll thing blinked.
This PR returns this behavior which is not good at all, but the problem is that it appears only when certain conditions are met.

Consider a random repo for an example:
https://github.com/Atom079/Designer

and the https://github.com/Atom079/Designer/blob/main/assets/sass/style.css file inside it.
All cursor-related settings I have are default (e.g. all cursors are shown, caret blinks):

https://github.com/zed-industries/zed/assets/2690773/5882a4e4-a663-42fe-9d23-9b4992e67c45

**Review by SomeoneToIgnore — APPROVED** [review]
Thank you for finding the culprit, clicking feels more consistent now: there is some lag, apparently if I start clicking quickly during/right after the scrollbar fading animation, but that looks like nitpicking already.

Let's merge it a bit later today after a new release, to try it on Nightly first, just in case.

File: crates/editor/src/element.rs [hunk 1]
```diff
@@ -2680,15 +2679,10 @@ impl EditorElement {
...
-                let point = cursor
-                    .0
-                    .to_display_point(&layout.position_map.snapshot.display_snapshot);
-                ColoredRange {
-                    start: point.row(),
-                    end: point.row(),
-                    color: cursor.1,
-                }
+            .map(|cursor| ColoredRange {
+                start: cursor.0.row(),
```
Comment (SomeoneToIgnore) [inline crates/editor/src/element.rs:2683]: NIT: could destructure `cursor` for readability.

**Comment by alygin** [comment]
> there is some lag, apparently if I start clicking quickly during/right after the scrollbar fading animation

Could you, please, elaborate a little bit on this? A lag between what events?

File: crates/editor/src/element.rs (hunk 1 above)
Comment (alygin) [reply crates/editor/src/element.rs]: Destructured

**Comment by SomeoneToIgnore** [comment]
Sorry, I do not have a way to record the video now and it's relatively rare, but if you try clicking back and forth the lines in the aforementioned file and lucky with the pattern, not all of your clicks will open the scrollbar, as intended.

I think it's minor and rare, and more concerned by the flashiness nature of the feature per se (not largely though too)

**Comment by alygin** [comment]
> not all of your clicks will open the scrollbar, as intended

Hm. If the problem is that the srollbar is not become visible when it's expected to, then I think the cause is not in the cursor markers. This is because the only case when they're expected to make the scrollbar visible, is when there's a cursor outside of the visible area. And it's not possible to make such a cursor by just clicking here an there without scrolling. At least I don't know how to do that. But scrolling makes the scrollbar become visible.

But I guess, I just don't understand your case, sorry.

I know one case in which the scrollbar is shown though it's not expected. One just needs to click somewhere in the currently visible area and make a small mouse movement while the button is still pressed (like in dragging). This small movement (if it's small enough) will not result in a text selection, yet it'll trigger `set_anchor()` which will call `show_scrollbar()`.

Remember that [bug with jumping scroll thumb](https://github.com/zed-industries/zed/pull/10667#issuecomment-2065059956)? This one looks very similar in how it's reproduced.

**Comment by SomeoneToIgnore** [comment]
https://github.com/zed-industries/zed/assets/2690773/8987ee7c-7727-4890-83c3-24cb45a760a9

I've tried to capture the slow flickering that bothered me, but the strange thing is that it becomes better when filmed (??).
Anyway, I clicked relatively similar most of the time, as the indicators show (in the late parts of the video, I did slow down intentionally to trigger the scroll fade, but that did not help to repro the bug much).
Yet, during clicking, around 10s and 13s and later, the indicator goes away despite me clicking more or less with the same rhythm — sometimes it is worse and longer, sometimes as brief and almost not notable as here.

Anyway, thank you for the global fix.

**Pull request merged**